Duplicated context concept guide (#34511)
quarkusio/quarkus: docs/src/main/asciidoc/duplicated-context.adoc

////
This document is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
[id="duplicated-context"]
= Duplicated context, context locals, asynchronous processing and propagation
include::_attributes.adoc[]
:diataxis-type: concept
:categories: core, architecture

> When using a traditional, blocking, and synchronous framework, each processing is separated in its own dedicated thread.

Reviewer: ```suggestion
When using a traditional, blocking, and synchronous framework, processing of each request is performed in a dedicated thread.
```

You know that this thread will not be used to execute anything else until the completion of the processing.
When you need to propagate data along the processing, such as the security principal or a trace id, you can use `ThreadLocals`.
The propagated data is cleared once the processing is complete.

When using a reactive and asynchronous execution model, you cannot use the same mechanism.
To avoid using many process threads, and reduce resource usage (and also increase the concurrency of the application), the same thread can be used to handle multiple concurrent processing.
Thus, you cannot use `ThreadLocals` as the values would be leaked between the various concurrent processing.

_Duplicated Contexts_ are a construct that provides the same kind of propagation but for asynchronous processing.
It can also be used with synchronous code.

This document explains duplicated contexts, how to retrieve one, use it, and how it's propagated along the (asynchronous) processing.

== The reactive model

NOTE: This section is not an explanation of the reactive model. Refer to that xref:quarkus-reactive-architecture.adoc[the Quarkus Reactive Architecture] for more details.

> Under the hood of Quarkus, there is a reactive engine.

Reviewer: ```suggestion
Under the hood, Quarkus uses a reactive engine.
```

image::reactive-continuation.png[alt=Continuation in the reactive execution model, width=80%, align=center]

You must **NEVER** block these event loops.
If you do, the whole model collapses.
Thus, when the processing of an HTTP request needs to execute an I/O operation (such as calling an external service), it:

1. schedules the operation,
2. passes a continuation (the code to call when the I/O completes),
3. releases the thread.

That thread can then handle another concurrent request.
When the scheduled operation completes, it executes the passed continuation **on the same event loop**.

That model is particularly efficient (low number of threads) and performant (avoid context switches).
However, it requires a different development model, and you cannot use _Thread Locals_, as the concurrent requests would see the same value.
Indeed, they are all processed by the same thread: the event loop.

The xref:context-propagation.adoc[MicroProfile Context Propagation] specification addresses this issue.
It saves and restores the values stored in thread locals whenever we switch to another processing unit.
However, that model is expensive.

> Context locals (also known as _duplicated context_) is another way to do this requiring fewer mechanics.

Reviewer: ```suggestion
Context locals (also known as _duplicated context_) is another way to do this and requires fewer mechanics.
```

=== Context locals restrictions

However, such a feature must only be used with duplicated contexts.
As said above, it's transparent for the code.
A duplicated context is a context, so they expose the same API.

In Quarkus, access to the local data is restricted to duplicated contexts.
If you try to access local data from a root context, it throws an `UnsupportedOperationException`.
It prevents accessing data shared among different units of processing.

[source, text]
----
java.lang.UnsupportedOperationException: Access to Context.putLocal(), Context.getLocal() and Context.removeLocal() are forbidden from a 'root' context  as it can leak data between unrelated processing. Make sure the method runs on a 'duplicated' (local) Context.
----

=== Safe context

You can mark a context _safe_.
It is meant for other extensions to integrate with to help identify which contexts are isolated and guarantee access by a unique thread.
Hibernate Reactive uses this feature to check if the current context is safe to store the currently opened session to protect users from mistakenly interleaving multiple reactive operations that could unintentionally share the same session.

> Vert.x web will create a new duplicated context for each http web request; Quarkus RestEasy Reactive will mark such contexts as safe.

Reviewer: ```suggestion
Vert.x web will create a new duplicated context for each http web request; Quarkus RESTEasy Reactive will mark such contexts as safe.
```
